nixos/modules/services/matrix/conduwuit.nix

```diff
@@ -0,0 +1,293 @@
+
+              Currently, conduwuit doesn't support inbound batched key requests, so
+              this list should only contain other Synapse servers.
+
+              Example: `[ "matrix.org" "constellatory.net" "tchncs.de" ]`
+            '';
+          };
+          global.database_path = lib.mkOption {
+            type = lib.types.path;
+            default = "/var/lib/conduwuit/";
+            description = ''
+              Path to the conduwuit database, the directory where conduwuit will save its data.
+              To change this value the DynamicUser feature has to be disabled for the conduwuit
+              systemd service.
+            '';
+          };
+          global.allow_check_for_updates = lib.mkOption {
+            type = lib.types.bool;
+            default = false;
+            description = ''
+              If enabled, conduwuit will send a simple GET request periodically to
+              <https://pupbrain.dev/check-for-updates/stable> for any new announcements made.
+              Despite the name, this is not an update check endpoint, it is simply an announcement check endpoint.
+
+              Disabled by default.
+            '';
+          };
+        };
+      };
+      default = { };
+      description = ''
+        Generates the conduwuit.toml configuration file. Refer to
+        <https://conduwuit.puppyirl.gay/configuration.html>
+        for details on supported values.
+        Note that database_path can not be edited because the service's reliance on systemd StateDir.
+      '';
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    assertions = [
+      {
+        assertion = cfg.settings.global.unix_socket_path == null || cfg.settings.global.address == null;
+        message = ''
+          In `services.conduwuit.settings.global`, `unix_socket_path` and `address` cannot be set at the
+          same time.
+          You can either listen on an IP address or a UNIX socket, but not both.
+          Leave one of the two options unset or explicitly set them to `null`.
+        '';
+      }
+      {
+        assertion =
+          config.systemd.services.conduwuit.serviceConfig.DynamicUser
+          -> cfg.settings.global.database_path == "/var/lib/conduwuit/";
+        message = ''
+          To change `services.conduwuit.settings.global.database_path`
+          `systemd.services.conduwuit.serviceConfig.DynamicUser` has to be disabled
+        '';
+      }
+    ];
+
+    # DynamicUser requires both user and group to exist or both not to exist.
+    # If we're configured to run under a different user or group, ensure
+    # the other default exists.
+    users =
+      lib.mkIf
+        (
+          cfg.settings.global.unix_socket_path == null || cfg.group != "conduwuit" || cfg.user != "conduwuit"
+        )
+        {
+          groups.${cfg.group} = lib.mkIf (cfg.group == "conduwuit") { };
+          users.${cfg.user} = lib.mkIf (cfg.user == "conduwuit") {
```

Comment:
This will always create the user and group since `lib.mkIf` returns an empty attrset if the expression is false.

Reply:
Good catch, thx. Another use case for https://github.com/NixOS/nixpkgs/pull/358646